[SYCL][Test] Replace multi-device kernel_bundle e2e tests with unit test

againull opened the pull request:
Replace the following e2e tests with unit test:
```
    ProgramManager/multi_device_bundle/device_libs_and_caching.cpp
    ProgramManager/multi_device_bundle/build_twice.cpp
    KernelAndProgram/persistent-cache-multi-device.cpp
```
    
The reason is that the checks performed in those e2e tests (number of UR calls, etc.) can be done more effificently using unit tests, such checks in e2e tests are expensive. Another reason is that e2e tests use NEO-specific `CreateMultipleRootDevices=[num]` environment variable which is not guaranteed to work, in unit tests we can emulate as many devices as we want.

againull commented:
@slawekptak @intel/llvm-reviewers-runtime Could you review please.

slawekptak reviewed (commented):
Just a small thing: This PR also modifies config.hpp. Could you please add a note to the description about this change and how it relates to the changes in tests?

againull commented:
> Just a small thing: This PR also modifies config.hpp. Could you please add a note to the description about this change and how it relates to the changes in tests?

Sorry, I've removed that part as I have realized it is excessive. Initially I have added specialization of the SYCLConfig for SYCL_DEVICELIB_NO_FALLBACK to be able to unset that config value from inside the unit test - because I want to test behavior when fallback libs are allowed. Now I've realized that there is a simpler way to do so using existing ScopedEnvVar and by passing nullptr to it (which results in unsetting the value, i.e. same desired behaviour).
